File: interface-anti-shake-demo/src/main/resources/lua/anti_duplicate_submit.lua
-- 1. Receive parameters: key=request identifier, timeout=expiration time
local key = KEYS[1]
local timeout = ARGV[1]

-- 2. Check lock: if exists return 1 (duplicate submission), if not exist set lock and return 0
if redis.call('EXISTS', key) == 1 then
    return 1
else
    redis.call('SET', key, '1', 'EX', timeout)
    return 0
end

File: interface-anti-shake-demo/src/main/java/com/example/antishake/annotation/AntiDuplicateSubmit.java
package com.example.antishake.annotation;

import java.lang.annotation.*;
import java.util.concurrent.TimeUnit;

@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
@Documented
public @interface AntiDuplicateSubmit {
    
    /**
     * Timeout for the lock (anti-shake time)
     * Default: 1 second
     */
    long timeout() default 1;
    
    /**
     * Time unit
     * Default: Seconds
     */
    TimeUnit unit() default TimeUnit.SECONDS;
    
    /**
     * Lock Strategy Type
     */
    LockType type() default LockType.REDIS_LUA;

    /**
     * Error message when duplicate submission is detected
     */
    String message() default "Request is too frequent, please try again later!";
    
    enum LockType {
        REDIS_LUA,
        REDISSON,
        LOCAL
    }
}
